src/Sketch.java

```diff
@@ -28,8 +67,73 @@ public void setup() {
      */
     public void draw() {
         background(180, 180, 255);
-        ball1.draw();
-        ball1.move();
+
+        for (Ball ball : ballsArray) {
+            ball.draw();
+            ball.move();
+        }
+
+        // for each of the 4 balls, check if their radius collides with another ball
+        // if it does, bounce the balls away from each other
+        for (int i = 0; i < ballsArray.length; i++) {
+            for (int j = i + 1; j < ballsArray.length; j++) {
+                if (ballsArray[i].collidesWith(ballsArray[j])) {
+                    ballsArray[i].bounce(ballsArray[j]);
+
+                    ballsArray[i].randomizeColors();
+                    ballsArray[j].randomizeColors();
+                }
+            }
+        }
+
+        for (int i = 0; i < ballsArray.length; i++) {
+            for (int j = 0; j < ballsArray.length; j++) {
+                if (ballsArray[i].collidesWithBubble(bubblesArray[j])) {
+                    bubblesArray[j].burst();
+                }
+            }
+        }
+
+        for (Bubble bubble : bubblesArray) {
+            bubble.draw();
+            bubble.move();
+        }
+
+        for (Snowflake snowflake : snowflakesArray) {
+            snowflake.draw();
+            snowflake.move();
+        }
+    }
+
+    /**
+     * Runs whenever the mouse is moved. Checks if cusor is on a ball to set it to
+     * frozen
+     */
+    public void mouseMoved() {
+        for (Ball ball : ballsArray) {
+            if (mouseX > ball.getX() - ball.getRadius() && mouseX < ball.getX() + ball.getRadius() && mouseY > ball.getY() - ball.getRadius() && mouseY < ball.getY() + ball.getRadius()) {
```

Comment:
using rectangular collision here instead of circular collision is a surprising choice. I think `if (distance(mouseX, mouseY, ball.getX(), ball.getY()) < ball.getRadius())` would have made more sense.